### PR TITLE
vcv-rack: 1.1.6 -> 2.0.6

### PR DESCRIPTION
**Pull request by jpotier**
###### Motivation for this change

Update to next major version, with great improvements in visual and sound engine, latency, usability, etc. https://vcvrack.com/Rack

It's a packaging of Rack 2 Free (the opensource one) + Fundamental (a default set of modules, like in VCV 1.x)

I've been using this for a couple of days for my own use, but I thought it may be a desirable update to have for everyone, so I'm opening this MR.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ddelabru** [comment]
When I first tried to build this with `nix-build -v -A vcv-rack` it failed because of hash mismatches. I was able to stop these mismatches by changing the sha256 sum for VCVRack/Fundamental to `1cnld5c682dfi6hkzs949zs05kranxgfivnzxszjd21ins10wz22` and the sha256 sum for VCVRack/Rack to `1xk7nrikrk7dnnk4ggjn1jis6qnywn52141mgixlcndml5na24la`. (Obtained these sums with `nix-prefetch-url --unpack "https://github.com/VCVRack/Rack/archive/v2.0.1.tar.gz"` etc.)

After that nix-build -v -A vcv-rack failed because the `prePatch` script line `cp dep/pffft/*.h dep/include` failed, because you removed the pffft source and other source dependencies from the derivation. I tried commenting out all these file copy operations that operate on sources that are no longer in the derivation, but the VCV Rack build still failed because it requires nanovg, and the derivation isn't supplying that in any way. How are you supplying these dependencies when you build VCV Rack 2 with this derivation?

**Comment by jpotier** [comment]
Hi @ddelabru, these dependencies are submodules of the vcv-rack git repository. These submodules are pulled (or at least they should be?) by `fetchFromGitHub`. `nix-prefetch-url --unpack` will justly give a different hash, because the source will be different.

It seems at least the bot was able to build it fine, so I'm not really sure what is happening here.

**Comment by jpotier** [comment]
Right, it seems that `fetchSubmodules` is not true by default, declared it explicitly.

**Comment by ddelabru** [comment]
Thanks for looking into this, @jpotier!

With `fetchSubmodules` turned on the build got much farther, but now is failing on a linker error:
```
g++ -o plugin.so build/src/8vert.cpp.o build/src/ADSR.cpp.o build/src/Delay.cpp.o build/src/LFO.cpp.o build/src/Merge.cpp.o build/src/MidSide.cpp.o build/src/Mixer.cpp.o build/src/Mutes.cpp.o build/src/Noise.cpp.o build/src/Octave.cpp.o build/src/Pulses.cpp.o build/src/Quantizer.cpp.o build/src/Random.cpp.o build/src/SEQ3.cpp.o build/src/Scope.cpp.o build/src/SequentialSwitch.cpp.o build/src/Split.cpp.o build/src/Sum.cpp.o build/src/Unity.cpp.o build/src/VCA-1.cpp.o build/src/VCA.cpp.o build/src/VCF.cpp.o build/src/VCMixer.cpp.o build/src/VCO.cpp.o build/src/Viz.cpp.o build/src/WTLFO.cpp.o build/src/WTVCO.cpp.o build/src/plugin.cpp.o build/src/dr_wav.c.o -shared -L../.. -lRack -Wl,-rpath=/tmp/Rack2 -static-libstdc++ -static-libgcc
/nix/store/lbxfixyw1yk099pjyaiy3xj5dl7kxm1g-binutils-2.35.2/bin/ld: cannot find -lRack
```

My other concern is that because these submodules don't specify a commit hash to pull in, the source hashes will be "fragile" and will no longer be valid when new commits are made to any of them. (I suspect this is why they had been specified directly in the derivation for Rack v1 to begin with.)

**Comment by jpotier** [comment]
> Thanks for looking into this, @jpotier!
> 
> With `fetchSubmodules` turned on the build got much farther, but now is failing on a linker error:
> 
> ```
> g++ -o plugin.so build/src/8vert.cpp.o build/src/ADSR.cpp.o build/src/Delay.cpp.o build/src/LFO.cpp.o build/src/Merge.cpp.o build/src/MidSide.cpp.o build/src/Mixer.cpp.o build/src/Mutes.cpp.o build/src/Noise.cpp.o build/src/Octave.cpp.o build/src/Pulses.cpp.o build/src/Quantizer.cpp.o build/src/Random.cpp.o build/src/SEQ3.cpp.o build/src/Scope.cpp.o build/src/SequentialSwitch.cpp.o build/src/Split.cpp.o build/src/Sum.cpp.o build/src/Unity.cpp.o build/src/VCA-1.cpp.o build/src/VCA.cpp.o build/src/VCF.cpp.o build/src/VCMixer.cpp.o build/src/VCO.cpp.o build/src/Viz.cpp.o build/src/WTLFO.cpp.o build/src/WTVCO.cpp.o build/src/plugin.cpp.o build/src/dr_wav.c.o -shared -L../.. -lRack -Wl,-rpath=/tmp/Rack2 -static-libstdc++ -static-libgcc
> /nix/store/lbxfixyw1yk099pjyaiy3xj5dl7kxm1g-binutils-2.35.2/bin/ld: cannot find -lRack
> ```

Hmm, it's breaking while building the Fundamental plugins pack. I'm not really able to tell why it's different for you than for me.

> My other concern is that because these submodules don't specify a commit hash to pull in, the source hashes will be "fragile" and will no longer be valid when new commits are made to any of them. (I suspect this is why they had been specified directly in the derivation for Rack v1 to begin with.)

I think you're right about this. What a chore! I'll be re-establishing separate git repos to pull the necessary dependencies, pinned.

**Comment by Ralith** [comment]
```
error: hash mismatch in fixed-output derivation '/nix/store/6fg7rfwsimhm4s62jfabx3f8xk449qyw-source.drv':
         specified: sha256-gy+YiYhgBO9eoEBP0PmteVjfaJWWIa/Ye9tvJG1b6aw=
            got:    sha256-ihKhbKG1WUZ7fDWQIIrl3mKjowxWvkemte3MPGO2Z/Y=
```

After fixing that:

```
patching sources
cp: missing destination file operand after 'dep/include'
Try 'cp --help' for more information.
```

How exactly are you testing this package?

**Comment by ddelabru** [comment]
@Ralith, the errors you reported look exactly like the errors I got when I first tried to build this; see my conversation with jpotier above. jpotier has amended the commit a few times and I no longer get the same errors with the latest code. (I'm getting a different error now, the linker error I described above.)

**Comment by jpotier** [comment]
> jpotier has amended the commit a few times and I no longer get the same errors with the latest code

Yeah, so I now pull the would-be submodules explicitly and separately from the git pull. Due to caching, I might have forgotten to revert the hash for the Rack source (grmbl) let me have a look again.

**Comment by Ralith** [comment]
> jpotier has amended the commit a few times and I no longer get the same errors with the latest code. (I'm getting a different error now, the linker error I described above.)

I am using the latest code. Check the hash for yourself: https://github.com/NixOS/nixpkgs/blob/656b21d8e0a91f60ca376941ddc75c00e2fdee3a/pkgs/applications/audio/vcv-rack/default.nix#L66

**Comment by jpotier** [comment]
@Ralith try it now? I've fixed the hash

**Comment by ddelabru** [comment]
I'm currently able to build with `nix-build -v -A vcv-rack` and the resulting binary runs. I should maybe make sure I don't have any cached sources that are bypassing hash mismatches.

**Comment by jpotier** [comment]
@ddelabru could you test that you are able:

1. to use the fundamental plugins (like VCV VCO, Scope, etc.)
2. to save a patch (I had to fix a bad crash due to some gtk stuff missing when I was building the package)

**Comment by ddelabru** [comment]
> @ddelabru could you test that you are able:
> 
>     1. to use the fundamental plugins (like VCV VCO, Scope, etc.)

The tutorial patch seems to include all these, and it works beautifully: the basic VCO, the wavetable VCO, scope, mixer, VCA, ADSR envelope generators, etc.

>     2. to save a patch (I had to fix a bad crash due to some gtk stuff missing when I was building the package)

Added the VCV Recorder module to the tutorial patch, saved it, and then closed and reopened Rack, and was able to load the saved patch without a problem.

**Comment by Ralith** [comment]
Can confirm, current version seems to work. Real-time output is super glitchy, though...

**Comment by jpotier** [comment]
> e: works properly when switched to pulse audio mode.

Good to know! I'm using PipeWire, and tried both Jack and Pulseaudio outputs from the Audio module of VCV, and both work pretty well.

**Comment by ddelabru** [comment]
> e: works properly when switched to pulse audio mode.

I always had stutter with ALSA output in Rack v1, too; switching to pulseaudio has always solved that unless I've so overloaded the GPU/CPU with modules that it's going to stutter anyway

**Comment by jpotier** [comment]
@Ralith @ddelabru would you mind doing a review of the code as well?

File: pkgs/applications/audio/vcv-rack/default.nix [hunk 1]
```diff
@@ -56,34 +75,40 @@ with lib; stdenv.mkDerivation rec {
...
+    chmod -R +rw plugins/Fundamental # will be used as build dir
+    substituteInPlace plugin.mk --replace ":= all" ":= dist"
   '';
 
   enableParallelBuilding = true;
 
-  nativeBuildInputs = [ makeWrapper pkg-config ];
-  buildInputs = [ alsa-lib curl glew glfw gtk2-x11 jansson libjack2 libsamplerate libzip rtaudio rtmidi speex ];
+  nativeBuildInputs = [ makeWrapper pkg-config wrapGAppsHook ];
+  buildInputs = [ alsa-lib curl ghc_filesystem glew glfw gtk3-x11 jansson jq libarchive libjack2 libpulseaudio libsamplerate rtaudio rtmidi speex zstd ];
```
Comment (Ralith) [inline pkgs/applications/audio/vcv-rack/default.nix:101]: Should `jq` be in `nativeBuildInputs`?

Comment (jpotier) [reply pkgs/applications/audio/vcv-rack/default.nix]: I move jq to nativeBuildInputs, good point.

**Comment by ddelabru** [comment]
@jpotier, the current iteration of the code looks good to me. However, there is now a 2.0.3 release. This one fixes a few major bugs people noticed when v2 was first made available.

**Comment by jpotier** [comment]
> @jpotier, the current iteration of the code looks good to me. However, there is now a 2.0.3 release. This one fixes a few major bugs people noticed when v2 was first made available.

I'll update to 2.0.3

**Comment by JoshuaFern** [comment]
Could we also keep v1 of vcv-rack? It has about 1,000+ more modules than VCV Rack 2 does at this time - some of which the authors have stated will never be updated.

**Comment by jpotier** [comment]
> Could we also keep v1 of vcv-rack? It has about 1,000+ more modules than VCV Rack 2 does at this time - some of which the authors have stated will never be updated.

Being able to still use VCV v1 is undeniably something important, and good to have.

NB: After this update, VCV v1 will still part of the history of Nixpkgs. As such, it will still be usable as long as the upstream source is available, albeit being a bit more difficult to discover. Now, I don't know if it is scheduled to be worked on independently of v2, or if it's scheduled to die upstream. If it is still patched for bugs upstream then yes, 100%, we should have separate packages for v1 and v2. However I was under the impression that v1 would be practically abandoned.

Personally, I don't mind extending the scope of this PR to create a new package in nixpkgs, and keep the v1 where it's at.

- Option 1: we move on and update v1 to v2, knowing it still allows people to run the older version, but that it's less transparent how to install v1 (need to pin an older nixpkgs), or
- Option 2: we create a new package for v2, to which the `vcv-rack` will default to, but there will still be a `vcv-rack1` package available to install v1.

@Ralith @ddelabru @Moredread @nathyong what do you people think about these two options?

**Comment by ddelabru** [comment]
I understand why people will want to keep Rack 1 around. There are certainly some very popular modules that either won't make it over to v2 or just haven't yet. That said, I also get the impression that v1 is being quickly abandoned by VCV. And, if I understand correctly, the VCV "Library" (essentially an app store that is the official place to obtain plugins and modules) is now hiding any plugins or modules that haven't been released for v2. So if a user doesn't already have the v1 plugins they want to use installed, they may have to go to the Github repos and web sites for those plugins and download them manually. For this reason, I think v2 should probably be the default `vcv-rack` package in nixpkgs 22.05, but maybe it should be a separate `vcv-rack-2` package in 21.11, with v1 remaining the default so that users' VCV Rack patches don't suddenly become unusable for them because they are running a version that doesn't have the modules they need.

**Comment by tholzschuh** [comment]
Yet another update has been released: the current version seems to be 2.0.4. @jpotier

**Comment by ddelabru** [comment]
I think they're up to 2.0.5 now; seems like they're still going after some bugfixes that canter up after the initial v2 release.

@tholzschuh Thanks for bringing up Rack Pro. I hadn't considered it but you're right that this derivation won't cover it. Patching a proprietary ELF binary to link in shared libraries from the Nix store isn't a problem for Nix derivations; that's how Nix packages the official Spotify client, for example. Since the free version can be built from source and the proprietary version would require patching an ELF binary, the pro version might need to be a separate package from the free version in Nix.

Right now I think the problem is that VCV Rack doesn't actually let you download the Rack Pro binary until you've bought it. (This is unlike other "freemium" software, for example the Bitwig DAW, where there's just one binary but at runtime it checks whether you've bought a license and opens either as the demo version or a paid version accordingly.) And I don't personally have Pro because Standalone Rack is sufficient for me, and money has been a little tight (though I have bought a commercial VCV plugin before and have my eye on others) so I can't test anything with the VCV Rack Pro binary. We should maybe reach out to the VST devs about how to enable Rack Pro on Nix / NixOS, and explain to them why that will require a patched binary, but it might help to have Nix users who have bought or would buy Rack Pro so we can demonstrate why they should support Nix compatibility.

The other wrinkle here is that the two distinguishing features of Pro are 1) professional support and 2) VCV Rack as a VST plugin for digital audio workstations (DAWs). So it only really makes sense to package VCV Rack Pro for Nix if we make sure the VST version works; I think someone who has more experience with VSTs should look at it.

Anyway, since it looks like Pro would have to be a separate package anyway, I think this update to the existing derivation should proceed, but maybe with a note that it's specifically for the free version of VCV Rack and not for Pro.

**Comment by tholzschuh** [comment]
> Right now I think the problem is that VCV Rack doesn't actually let you download the Rack Pro binary until you've bought it. (This is unlike other "freemium" software, for example the Bitwig DAW, where there's just one binary but at runtime it checks whether you've bought a license and opens either as the demo version or a paid version accordingly.) And I don't personally have Pro because Standalone Rack is sufficient for me, and money has been a little tight (though I have bought a commercial VCV plugin before and have my eye on others) so I can't test anything with the VCV Rack Pro binary. 

I bought the Pro version and would of course help where possible.
I'm not really experienced at all in packaging with nix (so far).

The Pro-version comes with a text file with installation instructions:
There's a Rack Pro folder that one should copy to some places, and two VST2 libraries that one should copy to some VST-locations.

> We should maybe reach out to the VST devs about how to enable Rack Pro on Nix / NixOS, and explain to them why that will require a patched binary, but it might help to have Nix users who have bought or would buy Rack Pro so we can demonstrate why they should support Nix compatibility.

As I said above: I'm a NixOS user that bought Rack Pro.
Maybe some others have too?
 
> The other wrinkle here is that the two distinguishing features of Pro are 1) professional support and 2) VCV Rack as a VST plugin for digital audio workstations (DAWs). So it only really makes sense to package VCV Rack Pro for Nix if we make sure the VST version works; I think someone who has more experience with VSTs should look at it.
> 
> Anyway, since it looks like Pro would have to be a separate package anyway, I think this update to the existing derivation should proceed, but maybe with a note that it's specifically for the free version of VCV Rack and not for Pro.

Yeah, I was just asking out of curiosity about the Pro-package.

The derivation suggested in this PR works perfectly well, so I'd definitely like to see it being accepted.

**Comment by jpotier** [comment]
bumped to v2.0.5

**Comment by ddelabru** [comment]
@jpotier, in 2.0.4 Rack switched from gtk3 to zenity for displaying dialog boxes, so I think you need to change the Nix derivation to reflect that. I can't save or open patches with the latest version, or do anything that requires dialogs.

**Comment by jpotier** [comment]
> in 2.0.4 Rack switched from gtk3 to zenity for displaying dialog boxes

Damn, really? Way to not abide by semver (that's on me for expecting a change from 2.0.4 -> 2.0.5 to be a patch thing, not a breaking change).

Thanks for the very detailed suggestion, @ddelabru I'll make the change as soon as I get some quiet time.

**Comment by ddelabru** [comment]
@tholzschuh, thanks for letting me know you have pro. Unfortunately I feel a little out of my depth making any attempt to package it if I don't have a copy of it myself but if you want to use Pro with Nix then there are probably others.

I've given a little more thought to how it should be done and it sounds like wha t happens when you download Pro after installing it is that you get some files (maybe in a zip archive or tarball?) that come with instructions on where to put the Pro executable, where to put your Rack directory, where to copy the VST, etc. Probably the best way to accommodate this, instead of adding Pro to nixpkgs, would be to maintain a simple Nix expression in a file that you would copy to your local directory alongside the file you downloaded from the VCV Rack web site, and then you'd use a Nix command like some variation of `nix-env -f ./ -iA vcv-rack-pro` to imperatively install Pro to your user profile. This wouldn't be declarative like nixpkgs but it would side-step the problems of needing to be authenticated with VCV Rack in order to download the Pro binary, and of unstable checksums that would result from grabbing possibly unversioned archive files. The Nix expression would take care of unpacking the archive, patching the binary, and putting all the files where they're supposed to go.

Hopefully some Nix packager who has a licence for Pro will have a go at this; if they do they could probably maintain it in their own repo since with the solution I'm suggesting it wouldn't belong in nixpkgs.

**Comment by ddelabru** [comment]
I had a look at the Rack source code and it seems the one line where the `zenity`  executable filename is defined isn't in the Rack source code itself but in a submodule https://github.com/AndrewBelt/osdialog.git , in the C source file `osdialog_zenity.c`:

```c
static const char zenityBin[] = "zenity";
```

So I think that's what needs patching specifically.

**Comment by jpotier** [comment]
@ddelabru there, should be fixed now? At least I could open the save menu with this new version.

**Comment by tasmo** [comment]
@jpotier I get now the following error:

```make
/nix/store/cdm6zywd51mbabxhklsixwcskv4n70s3-binutils-2.35.2/bin/ld: cannot find -lRack
collect2: error: ld returned 1 exit status
make[1]: *** [../../compile.mk:62: plugin.so] Error 1
make[1]: Leaving directory '/build/source/plugins/Fundamental'
make: *** [Makefile:280: plugins] Error 2
```

**Comment by ddelabru** [comment]
I was able to build commit 3a768c3e7f00866d48f1cdd5f5e23911cd987569 with `nix-build -A vcv-rack` and it produced a Rack binary 2.0.6 binary that worked well for me, with zenity dialogs and everything. Not sure why I didn't get the same linker error @tasmo did. @tasmo, were you building from the same git ref and with the same nix-build command?

I did try deleting the symlink to the build result and clearing my nix cache to force a fresh build, and it still worked.

**Comment by tasmo** [comment]
@ddelabru Yes, I used the same commit and the same command. Clearing the nix cache did it. thank you!

**Comment by Moredread** [comment]
Can you please remove me from the maintainer list with this PR (so that a seperate PR doesn't cause any merge conflicts)? I sadly have little time atm, and I don't think that will change soon.

**Comment by anna328p** [comment]
> [...] Right now I think the problem is that VCV Rack doesn't actually let you download the Rack Pro binary until you've bought it.

`requireFile`

**Comment by jpotier** [comment]
This has now been open for almost a quarter of a year, and I'm not feeling like progress is made.

I still have this random build problem other have expressed here too:

```
/nix/store/cdm6zywd51mbabxhklsixwcskv4n70s3-binutils-2.35.2/bin/ld: cannot find -lRack
```

Once in a blue moon.

I feel the PR is mostly good, but not there yet, and unfortunately, my itch was scratched (I now have a working version of Rack2 to use for myself, this one) so my motivation is mostly gone. I'll close my PR here, and I encourage anyone to pick it up and finish it for the community to enjoy, do not hesitate to reuse this PR and call it your own if it helps. I don't want to be the blocking factor in this process, and I feel that having an open PR for Rack2 in the pipes may prevent others to give it a go.

**Comment by yurkobb** [comment]
For anyone picking this up: [the license is now GPLv3 with plugin exception](https://github.com/VCVRack/Rack/blob/v2/CHANGELOG.md#100-2019-06-19) (meaning, plugins are allowed to be non-GPL if they aren't commercial) + cc by-nc-nd, the pr as of writing still has bsd. (This also affects the 1.x package).

**Pull request closed without merging**